lisp/tree-sitter.el

```diff
@@ -104,46 +121,76 @@ END is the end of the changed text."
             (ts-parse-chunks tree-sitter-parser #'ts-buffer-input tree-sitter-tree)))
     (run-hook-with-args 'tree-sitter-after-change-functions old-tree)))
 
-(defun tree-sitter--enable ()
+(defun tree-sitter--setup ()
   "Enable `tree-sitter' in the current buffer."
   (unless tree-sitter-language
     ;; Determine the language symbol based on `major-mode' .
     (let ((lang-symbol (alist-get major-mode tree-sitter-major-mode-language-alist)))
       (unless lang-symbol
-        ;; TODO: Consider doing nothing if the language is not supported, so
-        ;; that we can make this a global mode.
         (error "No language registered for major mode `%s'" major-mode))
       (setq tree-sitter-language (tree-sitter-require lang-symbol))))
   (unless tree-sitter-parser
     (setq tree-sitter-parser (ts-make-parser))
     (ts-set-language tree-sitter-parser tree-sitter-language))
-  (unless tree-sitter-tree
-    (tree-sitter--do-parse))
   (add-hook 'before-change-functions #'tree-sitter--before-change :append :local)
   (add-hook 'after-change-functions #'tree-sitter--after-change :append :local))
 
-(defun tree-sitter--disable ()
+(defun tree-sitter--teardown ()
   "Disable `tree-sitter' in the current buffer."
   (remove-hook 'after-change-functions #'tree-sitter--after-change :local)
   (remove-hook 'before-change-functions #'tree-sitter--before-change :local)
   (setq tree-sitter-tree nil
         tree-sitter-parser nil
         tree-sitter-language nil))
 
+(defmacro tree-sitter--error-protect (body-form &rest error-forms)
```

Comment:
Maybe you want to use `condition-case` instead?

Reply:
This doesn't handle the error. It runs some cleanup code and lets the error propagate.

The closest `condition-case` equivalent I can think of doesn't preserve the backtrace:

```emacs-lisp
`(condition-case err
    ,body-form
   (error
    ,@error-forms
    (signal (car err) (cdr err))))
```